### PR TITLE
Add support for instance monitoring

### PR DESCRIPTION
**Pull request by vendrov**
Add support for instance monitoring.
Please review this PR carefully since it's my first contribution.

#4482

### CODE REVIEW

**Comment by mikesplain** [comment]
This looks awesome! Let me take a look

/ok-to-test
/assign

**Comment by mikesplain** [comment]
Also, it'd be good to enable this in one of the tests rather than making it false. Complex is probably the best place:

https://github.com/kubernetes/kops/blob/96b6d1c6dc0b73b15ba6a3b643a7f6a899333f5f/tests/integration/update_cluster/complex/in-v1alpha2.yaml#L55-L76

File: docs/instance_groups.md [hunk 1]
```diff
@@ -308,3 +308,25 @@ spec:
...
+```
+# Example for nodes
+apiVersion: kops/v1alpha2
+kind: InstanceGroup
+metadata:
+  labels:
+    kops.k8s.io/cluster: k8s.dev.local
+  name: nodes
+spec:
+  DetailedInstanceMonitoring: true
```
Comment (mikesplain) [inline docs/instance_groups.md]: Should be `detailedInstanceMonitoring` to stay in line with our naming convention

File: pkg/apis/kops/instancegroup.go [hunk 2]
```diff
@@ -116,6 +116,8 @@ type InstanceGroupSpec struct {
 	AdditionalUserData []UserData `json:"additionalUserData,omitempty"`
 	// SuspendProcesses disables the listed Scaling Policies
 	SuspendProcesses []string `json:"suspendProcesses,omitempty"`
+	// DetailedInstanceMonitoring defines if detailed-monitoring is enabled (AWS only)
+	DetailedInstanceMonitoring *bool `json:"DetailedInstanceMonitoring,omitempty"`
```
Comment (mikesplain) [inline pkg/apis/kops/instancegroup.go]: `json:"detailedInstanceMonitoring,omitempty"`

File: pkg/apis/kops/v1alpha1/instancegroup.go [hunk 3]
```diff
@@ -96,6 +96,8 @@ type InstanceGroupSpec struct {
 	Zones []string `json:"zones,omitempty"`
 	// SuspendProcesses disables the listed Scaling Policies
 	SuspendProcesses []string `json:"suspendProcesses,omitempty"`
+	// DetailedInstanceMonitoring defines if detailed-monitoring is enabled (AWS only)
+	DetailedInstanceMonitoring *bool `json:"DetailedInstanceMonitoring,omitempty"`
```
Comment (mikesplain) [inline pkg/apis/kops/v1alpha1/instancegroup.go]: same as above

File: pkg/apis/kops/v1alpha1/zz_generated.conversion.go [hunk 4]
```diff
@@ -1748,6 +1748,7 @@ func autoConvert_v1alpha1_InstanceGroupSpec_To_kops_InstanceGroupSpec(in *Instan
 	}
 	out.Zones = in.Zones
 	out.SuspendProcesses = in.SuspendProcesses
+	out.DetailedInstanceMonitoring = in.DetailedInstanceMonitoring
```
Comment (mikesplain) [inline pkg/apis/kops/v1alpha1/zz_generated.conversion.go]: This will need to be regenerated again

File: pkg/apis/kops/v1alpha2/instancegroup.go [hunk 5]
```diff
@@ -105,6 +105,8 @@ type InstanceGroupSpec struct {
 	AdditionalUserData []UserData `json:"additionalUserData,omitempty"`
 	// SuspendProcesses disables the listed Scaling Policies
 	SuspendProcesses []string `json:"suspendProcesses,omitempty"`
+	// DetailedInstanceMonitoring defines if detailed-monitoring is enabled (AWS only)
+	DetailedInstanceMonitoring *bool `json:"DetailedInstanceMonitoring,omitempty"`
```
Comment (mikesplain) [inline pkg/apis/kops/v1alpha2/instancegroup.go]: Same as above.

File: pkg/apis/kops/instancegroup.go (hunk 2 above)
Comment (mikesplain) [reply pkg/apis/kops/instancegroup.go]: But don't change `DetailedInstanceMonitoring *bool`

**Comment by mikesplain** [comment]
This PR is making good progress, once we get some of those changes in and get the tests passing, I'll give it another test.

**Comment by mikesplain** [comment]
We're working offline on fixing this.

File: docs/instance_groups.md [hunk 6]
```diff
@@ -308,3 +308,25 @@ spec:
   suspendProcesses:
   - AZRebalance
 ```
+
+## Enabling Detailed-Monitoring on AWS instances
+
+Detailed-Monitoring will cause the monitoring data to be available every 1 minute instead of every 5 minutes. [Enabling Detailed Monitoring](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-cloudwatch-new.html). In production environments you may want to consider to enable detailed monitoring for quicker troubleshooting.
+
+**Note: that enabling detailed monitoring is a subject for [charge](https://aws.amazon.com/cloudwatch)**
```
Comment (mikesplain) [inline docs/instance_groups.md]: We can clean this up a little in a follow up PR but great first pass!

**Comment by justinsb** [comment]
/approve

Thanks @vendrov  - wonderful contribution!

I'm merging because it looks like terraform defaults to true (which is unusual), so this fixes a bug where terraform behaves differently.

**Pull request merged**